File: sdk-python/gatedhouse/events/adapters.py
"""Event bus adapters."""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Callable, Awaitable

from gatedhouse.core.types import GatehouseEvent

logger = logging.getLogger("gatedhouse.events.adapters")

EventHandler = Callable[[GatehouseEvent], Awaitable[None]]


class EventBusAdapter:
    """Base class for event bus adapters."""

    async def subscribe(self, topics: list[str], handler: EventHandler) -> None:
        raise NotImplementedError

    async def publish(self, topic: str, event: GatehouseEvent) -> None:
        raise NotImplementedError

    async def disconnect(self) -> None:
        raise NotImplementedError


class InMemoryEventBus(EventBusAdapter):
    """In-memory event bus for testing and development."""

    def __init__(self) -> None:
        self._handlers: list[EventHandler] = []

    async def subscribe(self, topics: list[str], handler: EventHandler) -> None:
        self._handlers.append(handler)

    async def publish(self, topic: str, event: GatehouseEvent) -> None:
        for handler in self._handlers:
            await handler(event)

    async def disconnect(self) -> None:
        self._handlers.clear()


class NoopEventBus(EventBusAdapter):
    """No-op event bus that silently discards events."""

    async def subscribe(self, topics: list[str], handler: EventHandler) -> None:
        pass

    async def publish(self, topic: str, event: GatehouseEvent) -> None:
        pass

    async def disconnect(self) -> None:
        pass

File: sdk-rust/src/events/adapters.rs
//! Event bus adapters.

use async_trait::async_trait;
use crate::types::GatehouseEvent;

#[async_trait]
pub trait EventBusAdapter: Send + Sync {
    async fn subscribe(&self, topics: &[String], handler: fn(&GatehouseEvent)) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
    async fn publish(&self, topic: &str, event: &GatehouseEvent) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
    async fn disconnect(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
}

/// In-memory event bus for testing and development.
pub struct InMemoryEventBus;

#[async_trait]
impl EventBusAdapter for InMemoryEventBus {
    async fn subscribe(&self, _topics: &[String], _handler: fn(&GatehouseEvent)) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }

    async fn publish(&self, _topic: &str, _event: &GatehouseEvent) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }

    async fn disconnect(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }
}

/// No-op event bus that silently discards events.
pub struct NoopEventBus;

#[async_trait]
impl EventBusAdapter for NoopEventBus {
    async fn subscribe(&self, _topics: &[String], _handler: fn(&GatehouseEvent)) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }

    async fn publish(&self, _topic: &str, _event: &GatehouseEvent) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }

    async fn disconnect(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Ok(())
    }
}
